### DISCUSSION TITLE
The first submission behaves differently than the following submissions when updating a form field

### DISCUSSION THREAD

**Discussion by christianledgard** [react-hook-form/react-hook-form#9210] [Q&A] [answered]
Hello!

I need to update some form values on the _validate_ statement, but my program is not behaving correctly. When I submit a form for the first time, the value is not updated, but in the following submissions, the value is updated correctly. I simulate the issue on a [CodeSandbox](https://codesandbox.io/s/setvalue-is-not-working-at-the-first-submit-5w5r50?file=/src/index.js). 

As we see in the image, the first console log reproduces only the field value, and on the second submission, both values are displayed. My intended behavior is to obtain a `{example: "test", example2: true}` on the first submission.

<img width="1463" alt="CodeSandbox screenshot..." src="https://user-images.githubusercontent.com/39962750/195452985-adabc02b-84af-4be6-ad4c-5b06c48485bf.png">

For reproducing it:
1. [Code on CodeSandbox](https://codesandbox.io/s/setvalue-is-not-working-at-the-first-submit-5w5r50?file=/src/index.js).
2. Open the CodeSandbox console.
3. Submit multiple forms with random text. 

Thank you for the help!

**Comment by leapful** [comment, marked as answer]
@christianledgard : It seems you are mis-understanding how validation works in react-hook-form and also the normal behavior of submitting form. 

Form values should be unchanged while validating, you should not update form values on validation or submission, it's a best practice when dealing with form as any change on the fly can cause unexpected issue on other field values.

Method "setValue" has similar behavior like React set state, it's an async handling therefore when calling "setValue" in validate function, the new value will be only available on the next re-rendering, not when submitting form.

If you want to update data that depends on validation result, just perform validation after success callback of method "handleSubmit" is triggered.

For example:

```
const onSubmit = handleSubmit((data) => {
    // Validate something
    ...
    // Update data
    data.example2 = "Some values"

   // then submit the whole form values
   ...
})
```

**Reply by christianledgard** [reply]
Thank you @leapful for your time and clear explanation. I just transferred the logic to `handleSubmit` and it is working as expected. 👍🏻
